### PR TITLE
Feature/saas 9853 aws cognito user pool have waf

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            }
+            for (let userPool of userPools.data) {
+                var webACLResource = helpers.addSource(cache, source,
+                    ['wafv2', 'getWebACLForResource', region, userPool.Id]);
+              
+                if (!webACLResource || webACLResource.err || !webACLResource.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for wafv2 api: ' + helpers.addError(webACLResource), region);
+
+                    return rcb();
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:46]:
```suggestion
                    continue;
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:46]: use continue in for loop

File: plugins/aws/cognito/cognitoHasWafEnabled.spec.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                'us-east-1': {
+                    err: null,
+                    data: poolList
+                }
+            },
+        },
+        sts: {
+            getCallerIdentity: {
+                'us-east-1': {
+                data: '101363884315'
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.spec.js:25]:
```suggestion
                data: '11111222222'
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.spec.js:25]: use dummy id

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
+                var webACLResource = helpers.addSource(cache, source,
+                    ['wafv2', 'getWebACLForResource', region, userPool.Id]);
+              
+                if (!webACLResource || webACLResource.err || !webACLResource.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for wafv2 api: ' + helpers.addError(webACLResource), region);
+
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:44]:
```suggestion
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
+                var webACLResource = helpers.addSource(cache, source,
+                    ['wafv2', 'getWebACLForResource', region, userPool.Id]);
+              
+                if (!webACLResource || webACLResource.err || !webACLResource.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for wafv2 api: ' + helpers.addError(webACLResource), region);
+                    return rcb();
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:44]:
```suggestion
```suggestion
                    continue;
```
```

Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:44]:
```suggestion
                    continue;
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:44]: use continue in for loop

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query api: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No User pool found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:37]:
```suggestion
            for (let userPool of userPools.data) {
                  if(!userPool.Id) continue;
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:37]: add check to make sure userPool.Id is present

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'EKS Cluster Has Tags',
+    category: 'EKS',
+    domain: 'Containers',
+    description: 'Ensure that AWS EKS Clusters have tags associated.',
+    more_info: 'Tags help you to group resources together that are related to or associated with each other. It is a best practice to tag cloud resources to better organize and gain visibility into their usage.',
+    link: 'https://docs.aws.amazon.com/eks/latest/userguide/eks-using-tags.html',
+    recommended_action: 'Modify EKS Cluster and add tags.',
```
[the remark this answers is not in the document]
Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:11]: it should be according to cognito WAF

File: plugins/aws/cognito/cognitoHasWafEnabled.js (hunk 5 above)
Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:37]: add check to make sure userPool.Id is present

File: plugins/aws/cognito/cognitoHasWafEnabled.js (hunk 4 above)
Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.js:44]: use continue in for loop

File: plugins/aws/cognito/cognitoHasWafEnabled.spec.js (hunk 2 above)
Comment (alphadev4) [reply plugins/aws/cognito/cognitoHasWafEnabled.spec.js:25]: use dummy id

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'COgnito UserPool Has WAF Enabled',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:5]:
```suggestion
    title: 'Cognito User Pool Has WAF Enabled',
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has WAF Enabled',
+    category: 'CognitoIdentityServiceProvider',
+    domain: 'Identity Service Provider',
+    description: 'Ensure that Cognito UserPool has WAF enabled.',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:8]:
```suggestion
    description: 'Ensure that Cognito User Pool has WAF enabled.',
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                
+                if (!userPool.Id) continue;
+
+                var webACLResource = helpers.addSource(cache, source,
+                    ['wafv2', 'getWebACLForResource', region, userPool.Id]);
+              
+                if (!webACLResource || webACLResource.err || !webACLResource.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for wafv2 api: ' + helpers.addError(webACLResource), region);
+
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:47]:
```suggestion
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query api: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No User pool found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
+
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:38]:
```suggestion
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has WAF Enabled',
+    category: 'CognitoIdentityServiceProvider',
+    domain: 'Identity Service Provider',
+    description: 'Ensure that Cognito User Pool has WAF enabled.',
+    more_info: 'Enabling WAF allows control over requests to the load balancer, allowing or denying traffic based off rules in the Web ACL.',
+    link: 'https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-waf.html',
+    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable WAF from properties ',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:11]:
```suggestion
    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable WAF from properties',
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has WAF Enabled',
+    category: 'CognitoIdentityServiceProvider',
+    domain: 'Identity Service Provider',
+    description: 'Ensure that Cognito User Pool has WAF enabled.',
+    more_info: 'Enabling WAF allows control over requests to the load balancer, allowing or denying traffic based off rules in the Web ACL.',
+    link: 'https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-waf.html',
+    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable WAF from properties',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:11]:
```suggestion
    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable WAF from properties.',
```

File: helpers/aws/api.js [hunk 13]
```diff
@@ -2061,7 +2061,13 @@ var postcalls = [
                 reliesOnService: 'wafv2',
                 reliesOnCall: 'listWebACLs',
                 override: true
+            },
+            getWebACLForCognitoUserPool: {
+                reliesOnService: 'cognitoidentityserviceprovider',
+                reliesOnCall: 'listUserPools',
+                override: true
             }
+
```
Comment (AkhtarAmir) [inline helpers/aws/api.js:2070]:
```suggestion
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query Cognito user pools: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No Cognito user pools found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:37]:
```suggestion
            }

            if (!userPools.data.length){
                helpers.addResult(results, 0, 'No Cognito user pools found', region);
                return rcb();
            }

            for (let userPool of userPools.data) {
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+              
+                if (!webACLResource || webACLResource.err || !webACLResource.data){
+                    helpers.addResult(results, 3,
+                        'Unable to get WebACL resource for cognito user pool: ' + helpers.addError(webACLResource), region, arn);
+                    continue;
+                }
+                if (webACLResource.data.WebACL){
+                    helpers.addResult(results, 0, 'Cognito User pool has WAFV2 enabled', region, arn);
+                } else {
+                    helpers.addResult(results, 2, 'Cognito User pool does not have WAFV2 enabled', region, arn);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:53]:
```suggestion
                    helpers.addResult(results, 0, 'User pool has WAFV2 enabled', region, arn);
                } else {
                    helpers.addResult(results, 2, 'User pool does not have WAFV2 enabled', region, arn);
```

File: plugins/aws/cognito/cognitoHasWafEnabled.spec.js [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            const cache = createCache([listUserPools[0]], null);
+            cognitoHasWafEnabled.run(cache, {}, callback);
+        });
+
+        it('should give passing result if User pool has WAF enabled', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Cognito User pool has WAFV2 enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoHasWafEnabled.spec.js:88]:
```suggestion
                expect(results[0].message).to.include('User pool has WAFV2 enabled');
```

File: plugins/aws/cognito/cognitoHasWafEnabled.spec.js [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+            const cache = createCache([listUserPools[0]], { WebACL: {'Name': 'abc'}});
+            cognitoHasWafEnabled.run(cache, {}, callback);
+        });
+
+        it('should give failing result if User pool does not have WAF enabled', function (done) {
+                const callback = (err, results) => {
+                    expect(results.length).to.equal(1);
+                    expect(results[0].status).to.equal(2);
+                    expect(results[0].region).to.equal('us-east-1');
+                    expect(results[0].message).to.include('Cognito User pool does not have WAFV2 enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoHasWafEnabled.spec.js:101]:
```suggestion
                    expect(results[0].message).to.include('User pool does not have WAFV2 enabled');
```

File: plugins/aws/cognito/cognitoHasWafEnabled.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has WAF Enabled',
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoHasWafEnabled.js:5]:
```suggestion
    title: 'Cognito User Pool WAF Enabled',
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**